libs/header/sync/sync.go

```diff
@@ -299,32 +292,32 @@ func (s *Syncer[H]) requestHeaders(
 		if amount < size {
 			size = amount
 		}
-		headers, err := s.exchange.GetVerifiedRange(ctx, fromHead, size)
+
+		headers, err := s.getter.GetVerifiedRange(ctx, fromHead, size)
 		if err != nil {
 			return err
 		}
-		amount -= size
+		// TODO(@Wondertan): This fixed syncing for me, but the len should always be = to size
+		//  Something fishy in here.
```

Comment:
😱

Reply:
can you open issue for this at least

Reply:
I'll actually retry again if it's really the case, as things changed since then. Thanks for tagging this, as I completely forgot to double check.

Reply:
Have you observed it on the network or in tests also?

Reply:
I retried, and on the second attempt syncing broke. It went requesting things beyond network head
![image](https://user-images.githubusercontent.com/22449364/225008485-dbf6a768-f990-4d0a-96b2-9dc59ef97a63.png)

Reply:
Yes. The same for me. And the root cause is inside `ranges`. For some reason, it stores already handled header. I run node on your branch with extra logs and it shows that:
```
// finished requesting the range
2023-03-14T15:36:40.006+0200	DEBUG	header/p2p	p2p/session.go:109	received headers range	{"from": 212994, "to": 213336}
```

```
// applied cached head
2023-03-14T15:36:40.050+0200	INFO	header/store	store/store.go:349	new head	{"height": 213337, "hash": "566CE3183AD172A408FEA9FA7FE60648C8BEF6E153266EFB8913CF69FE7BFB2D"}
```

```
// and then got the same header from the cache after 
		headers := headersRange.Truncate(to) // amount=1 // 3066
		if len(headers) == 0 {
			break
		}
```
 
`{"header":{"version":{"block":"11"},"chain_id":"arabica-6","height":"213337","time":"2023-03-14T13:34:12.03287407Z"`

Reply:
We found the root cause and its now fixed. It could happen in case we receive a duplicate from pubsub

Reply:
Even though I had this case in mind while making this PR, I totally forgot that we truncated the pending just right when we started syncing, and @vgonkivs proposed to split Truncate into Get and Remove, so we clean up pending only after processing a range finished and not before. This way, pending will keep the sync target long enough, and duplicates will be detected/ignored correctly.